In [1]:
import pandas as pd
from pathlib import Path
import os
import plotly.express as px

In [2]:
absolute_path = Path(os.path.abspath('EDA.ipynb')).parent.parent
df_path = absolute_path.joinpath('vehicles_us.csv')
df = pd.read_csv(df_path)


In [3]:
df.head(5)

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [5]:
df.isna().sum()

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

In [6]:
df.describe()

price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  47906.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.750470      6.125235  115553.461738      1.0   
std     10040.803015      6.282065      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2006.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2014.000000      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

In [7]:
df.is_4wd.value_counts(dropna=False)

is_4wd
NaN    25953
1.0    25572
Name: count, dtype: int64

This column refers to is car 4 wd. There is only 1 and NaN, so we can define that 1 is True and NaN is False.

In [8]:
df.is_4wd = df.is_4wd.fillna(0)
df.is_4wd = df.is_4wd.astype(bool)

In column that define color all NaN shoul be changed to unknown.

In [9]:
df.paint_color = df.paint_color.fillna('unknown')

In [10]:
df.cylinders = df.cylinders.fillna('unknown')

In [11]:
df.model_year.unique()

array([2011.,   nan, 2013., 2003., 2017., 2014., 2015., 2012., 2008.,
       2018., 2009., 2010., 2007., 2004., 2005., 2001., 2006., 1966.,
       1994., 2019., 2000., 2016., 1993., 1999., 1997., 2002., 1981.,
       1995., 1996., 1975., 1998., 1985., 1977., 1987., 1974., 1990.,
       1992., 1991., 1972., 1967., 1988., 1969., 1989., 1978., 1965.,
       1979., 1968., 1986., 1980., 1964., 1963., 1984., 1982., 1973.,
       1970., 1955., 1971., 1976., 1983., 1954., 1962., 1948., 1960.,
       1908., 1961., 1936., 1949., 1958., 1929.])

In [12]:
model_year_df = df.groupby(by= 'model_year')['price'].agg(['mean', 'count'])

mean  count
model_year                
1908.0      12995.0      2
1929.0      18000.0      1
1936.0       5000.0      1
1948.0      21000.0      1
1949.0      44900.0      1

In [13]:
fig = px.scatter(model_year_df, y = 'mean', trendline = 'lowess')
fig.update_layout(
    xaxis_title = 'model year',
    yaxis_title = 'price'
)
fig.show()

There is some non-linear trendline here between price and  model year. Unfortunatly in dataset we have 3619 NaN values and for proper analysis in future we have to get rid of them.

In [14]:
fig = px.scatter(model_year_df, y = 'count')
fig.update_layout(
    xaxis_title = 'model year',
    yaxis_title = 'quantity'
)
fig.show()

 Dropping all NaN values in model year

In [15]:
df.head()

price  model_year           model  condition cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good       6.0  gas  145000.0   
1  25500         NaN      ford f-150       good       6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new       4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair       8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent       4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV     unknown    True  2018-06-23           19  
1    automatic  pickup       white    True  2018-10-19           50  
2    automatic   sedan         red   False  2019-02-07           79  
3    automatic  pickup     unknown   False  2019-03-22            9  
4    automatic   sedan       black   False  2019-04-02           28

In [16]:
df['odometer_adjusted'] = round(df['odometer']/1000)
odometer_df = df.groupby('odometer_adjusted')['price'].mean()


In [17]:
fig = px.scatter(odometer_df)
fig.update_layout(
    xaxis_title = 'odometer',
    yaxis_title = 'mean price',
    xaxis_ticksuffix = 'K'
)
fig.show()

In [18]:
df.model.str.split(' ', n = 1, expand = True)

0        1
0           bmw       x5
1          ford    f-150
2       hyundai   sonata
3          ford    f-150
4      chrysler      200
...         ...      ...
51520    nissan   maxima
51521     honda    civic
51522   hyundai   sonata
51523    toyota  corolla
51524    nissan   altima

[51525 rows x 2 columns]

In [21]:
df[['manufacture', 'model']] = df.model.str.split(' ', n = 1, expand = True)

In [23]:
df.manufacture.unique()

array(['bmw', 'ford', 'hyundai', 'chrysler', 'toyota', 'honda', 'kia',
       'chevrolet', 'ram', 'gmc', 'jeep', 'nissan', 'subaru', 'dodge',
       'mercedes-benz', 'acura', 'cadillac', 'volkswagen', 'buick'],
      dtype=object)

In [44]:
df_manufacturer = df.groupby(['manufacture', 'type'])['type'].count().unstack(fill_value=0)
fig = px.bar(df_manufacturer)
fig.show()
# df_manufacturer.head(10)

In [58]:
df_condition = df.groupby(['model_year', 'condition'])['condition'].count()
fig = px.histogram(df_manufacturer)
fig.show()
# df_condition.head()